### PR TITLE
gnome: Fix the issue `Using the 'memory' GSettings backend issue`

### LINKED ISSUE #25906: Gnome 3 doesn't remember favorites
## Issue description
In Gnome 3 you can select applications as "favorites" to remain pinned to the sidebar. Under NixOS every time I reboot I lose my favorites.

### Steps to reproduce

Install NixOS with Gnome.

## Technical details

* System: (NixOS: `nixos-version`, Ubuntu/Fedora: `lsb_release -a`, ...)
17.09pre106916.c5badb123a (Hummingbird)
* Nix version: (run `nix-env --version`)
nix-env (Nix) 1.11.9
* Nixpkgs version: (run `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion`)
"17.09pre107265.0afb6d789c"
* Sandboxing enabled: (run `grep build-use-sandbox /etc/nix/nix.conf`)
No

### PR DESCRIPTION
**Pull request by midchildan**
###### Motivation for this change
Fixes #25906

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by 0xABAB — APPROVED** [review]

**Comment by eqyiel** [comment]
It looks like this is all good but the Travis job was terminated because of the massive rebuild, thanks for fixing this!

**Comment by jtojnar** [comment]
Looks good but I still see

```
telepathy-idle[2133]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
```

File: pkgs/desktops/gnome-3/3.22/core/gnome-control-center/default.nix [hunk 1]
```diff
@@ -26,7 +26,7 @@ stdenv.mkDerivation rec {
       accountsservice libkrb5 networkmanagerapplet libwacom samba libnotify libxkbfile
       shared_mime_info icu libtool docbook_xsl docbook_xsl_ns gnome3.grilo
       gdk_pixbuf gnome3.defaultIconTheme librsvg clutter clutter_gtk
-      gnome3.vino udev libcanberra_gtk3 libgudev
+      gnome3.dconf gnome3.vino udev libcanberra_gtk3 libgudev
       networkmanager modemmanager makeWrapper gnome3.gnome-bluetooth grilo tracker ];
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/3.22/core/gnome-control-center/default.nix]: `makeWrapper` has to be replaced with `wrapGAppsHook` for control center to work.

**Comment by grahamc** [comment]
@midchildan How would I test this best?

**Comment by jtojnar** [comment]
@grahamc Something like

```bash
cd nixpkgs
git remote add midchildan git@github.com:midchildan/nixpkgs.git
git fetch midchildan
git checkout midchildan/fix-gnome
sudo nixos-rebuild switch -I nixpkgs=.
```

**Comment by midchildan** [comment]
@grahamc @jtojnar AFAIK, that seems to be the only way.

**Comment by eqyiel** [comment]
Hey @midchildan, I've been running this for a couple of days without issue, but I'm not sure what kind of results I should be seeing.

Do you remember what issue this resolved for you so that I can try it too?

I'm still seeing a lot of this in the journal:

```
$ journalctl -b 0 | grep -i gsettings
Jul 23 09:53:13 ayanami gsettings[1152]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:15 ayanami .gnome-shell-wr[1197]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:16 ayanami .evolution-sour[1229]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:16 ayanami .mission-contro[1250]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:17 ayanami .gnome-shell-ca[1225]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:19 ayanami .evolution-cale[1359]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:21 ayanami ..syncthing-gtk[1328]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:21 ayanami .electron-wrapp[1335]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:22 ayanami .evolution-cale[1558]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:22 ayanami .evolution-addr[1571]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:23 ayanami .evolution-addr[1610]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:32 ayanami .gnome-contacts[1674]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:32 ayanami .evolution-addr[1829]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:36 ayanami firefox[1912]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
Jul 23 09:53:57 ayanami Web Content[2029]: Using the 'memory' GSettings backend.  Your settings will not be saved or shared with other applications.
```

And things like changing the wallpaper have no effect (that was an issue before this PR though 😝 ).

**Comment by midchildan** [comment]
@eqyiel I only fixed the issue for packages that I noticed had a missing dependency on dconf. This issue have to be fixed on a per-package basis, so I guess there are still a lot of packages out there that also needs to be fixed.

**Comment by bjornfor** [comment]
hamster-time-tracker also need this, I think. (Note to self.)

**Comment by jtojnar** [comment]
I was trying to fix `gnome-system-monitor` and adding the dconf GIO variable did not seem to help. On the other hand just a simple `wrapGAppsHook` did work.

Additionally, the wrapping should specify a separator:

`--prefix GIO_EXTRA_MODULES : "${dconf}/lib/gio/modules"` instead of `--prefix GIO_EXTRA_MODULES "${dconf}/lib/gio/modules"`

**Comment by jtojnar** [comment]
Adding `${gnome3.dconf}/lib/gio/modules` points `GIO_EXTRA_MODULES` to an inexistent directory `/nix/store/fnwqja5vshafjdlmwc1c54h4pqklrsq0-dconf-0.26.0/lib/gio/modules`. On the other hand `wrapGAppsHook` links to `/nix/store/7snbgaz22204wpril41sr9prnqqg5636-dconf-0.26.0-lib/lib/gio/modules` (the `lib` output). It looks like #24133 is to blame.

The solution would be to use `gnome3.dconf.lib` instead of `gnome3.dconf` or use `wrapGAppsHook`, which already depends on the `lib` output.

**Comment by jtojnar** [comment]
I added some more fixes and rebased https://github.com/jtojnar/nixpkgs/tree/fix-gnome

**Comment by eqyiel** [comment]
@jtojnar thanks for this, I've been running your branch for a while now and this command produces no output:

```
% journalctl -b 0 | grep -i gsettings; echo $?
1 
```

On top of that, it resolves a bunch of other issues I've had with Gnome on the unstable branch for the last few months (keyboard/libinput/libinput settings are respected).

**Comment by jtojnar** [comment]
@midchildan Could you merge my commit so we could proceed?

**Comment by midchildan** [comment]
@jtojnar Merged :-)

**Comment by jtojnar** [comment]
It should be ready to merge, pinging GNOME maintainers @lethalman and @jgeerds.

**Comment by jgeerds** [comment]
@jtojnar sorry but I'm no longer a maintainer of Gnome.

**Pull request merged**